### PR TITLE
fixed a resource leak that prevented status iterators from cleaning properly

### LINKED ISSUE #134: Dangling async iterator after closing connection
### Observed behavior

Using (test) code below to monitor the status of a connection:

```
const lock = Lock(3);
const { nc } = await setup({}, { reconnect: false });

(async () => {
  for await (const s of nc.status()) {
    switch (s.type) {
      case Events.Disconnect:
        lock.unlock();
        break;
    }
  }

  lock.unlock()
})().then();

nc.closed().then(() => {
  lock.unlock();
});

await nc.close();
await lock;
```

causes a dangling async iterator to endlessly await the "submission" of a new signal, as the iterator instance in the `NatsConnectionImpl` instance is *not* stopped upon calling `close`. This causes memory leaks when connections are (intentionally) closed and re-created.

### Expected behavior

The async iterator created by invoking `status` on the `NatsConnection` is stopped and doesn't wait for the "submission" of a new signal, after the connection is closed explicitly.

### Server and client version

@nats-io/transport-node: 3.0.0-20

### Host environment

Node.js: 20.11.1

### Steps to reproduce

_No response_

### PR DESCRIPTION
**Pull request by aricart**
Fix #134

### CODE REVIEW

**Review by scottf — APPROVED** [review]
LGTM

File: core/src/protocol.ts [hunk 1]
```diff
@@ -1037,19 +1034,24 @@ export class ProtocolHandler implements Dispatcher<ParserEvent> {
...
+    const proms = [];
+    for (let i = 0; i < this.listeners.length; i++) {
+      const qi = this.listeners[i];
+      if (qi) {
+        qi.stop();
+        proms.push(qi.iterClosed);
+      }
+    }
+    if (proms.length) {
+      await Promise.all(proms);
```
Comment (mtmk) [inline core/src/protocol.ts:1046]: what happens if there was multiple proms and one threw an error?

Comment (aricart) [reply core/src/protocol.ts:1046]: those don't throw - the could resolve to an error but they don't reject

**Review by scottf — APPROVED** [review]
LGTM

**Review by mtmk — APPROVED** [review]
LGTM

**Review by scottf — APPROVED** [review]
LGTM

**Pull request merged**